### ISSUE TITLE
MarcSpec for Python pymarc

### ISSUE THREAD

**Issue by danizen** [MARCspec/MARCspec#29]
@cKlee, are you still maintaining this organization and the several implementations?

**Comment by cKlee** [comment]
Yes I do.

**Comment by danizen** [comment]
OK - I am then working on an implementation on top of pymarc and will let you know how it goes.  I will probably do it as part of National Library of Medicine's implementation of Alma, but then try to release it as a separate package and let you know.

Before I close this, I have a question of compliance relevant to the Python compiler-builder TATSU.   Would 005[1-5] be legal with spaces added like 005 [ 1-5 ] or not?

**Comment by danizen** [comment]
The reason I ask is that it looks like I need to make each rule fully upper case to prevent skipping whitespace:

```python
In [5]: parser.parse('005[1-7]')
Out[5]: ['005', '[', ['1', '-', '7'], ']', []]

In [6]: parser.parse('005 [1-7]')
Out[6]: ['005', '[', ['1', '-', '7'], ']', []]
```

I will of course determine validity before I add a model or fuller AST.

**Comment by danizen** [comment]
@edsu and @cKlee - welcoming help and not sure where it will ultimately go - https://github.com/danizen/pymarcspec/blob/main/marcspec/marcspec.ebnf

I publish now because I am stuck.  I am using tatsu, which is a PEG Parser.   So, order of productions matters.  I have re-ordered enough to get basic fieldSpec, subfieldSpec and indicatorSpec rules working, but I am having some trouble with comparisonString - to me the rule seems non-terminating, because /[:print]/ is very broad match - we want '}' to terminate, and that will not happen with a PEG Parser which consumes the '}'.

I have chosen for now to diverge in subfieldChar, but will fix it eventually.

I am on a very tight schedule and so I put this out knowing that what I have at work is going to diverge and if all goes well, I will have to switch to the version in pymarc or in my own package later on...

**Comment by cKlee** [comment]
> Before I close this, I have a question of compliance relevant to the Python compiler-builder TATSU. Would 005[1-5] be legal with spaces added like 005 [ 1-5 ] or not?

No, whitespaces are definitely not allowed. Feels like ages I worked with the spec. I thought about allowing the often seen '245 $a'. But I dropped it for reducing complexity, I think.

**Comment by cKlee** [comment]
> I publish now because I am stuck. I am using tatsu, which is a PEG Parser. So, order of productions matters. I have re-ordered enough to get basic fieldSpec, subfieldSpec and indicatorSpec rules working, but I am having some trouble with comparisonString - to me the rule seems non-terminating, because /[:print]/ is very broad match - we want '}' to terminate, and that will not happen with a PEG Parser which consumes the '}'.

I see. The PEG Parser uses the ebnf rules. And the problem is that '}' should end parsing the subspec. But the use /[:print]/ entails '}' and thus leeds to the problem?

Something like ``{$a~\abc\}def}`` should fail then even if it is properly escaped, right?

**Comment by danizen** [comment]
> I see. The PEG Parser uses the ebnf rules. And the problem is that '}' should end parsing the subspec. But the use /[:print]/ entails '}' and thus leeds to the problem?

Not quite - an honest translation of the ABNF rule:

```
comparisonString  = "\" *VCHAR
```

to tatsu EBNF:

```
comparisonString  = "\" {VCHAR} ;
```

The problem is that `/[:print]/` will match the closing `}` of the surrounding subSpec, and that causes the string not to be matched.  If I use instead this rule:

```
comparisonString  = "\" ~ /[a-z0-9A-Z?\-_]*/ ;
```

Then it works, but it is no longer accurate to MarcSpec.  It may get me through my project since I likely only need compSpec to match only if the 1st or 2nd indicator is a particular value, but it isn't good enough for you guys to simply include without some notes.

So, I am now unstuck, but still blocked from creating a proper implementation :)

**Comment by danizen** [comment]
Another question - why not a unaryOperator and binaryOperator rule, e.g. why are `?` and `!` made binary operators anyway:

```
operator          = "=" / "!=" / "~" / "!~" / "!" / "?"
                    ; equal / unequal / includes / not includes / not exists / exists
```

**Comment by danizen** [comment]
On the previous issue, I broadened the vchar rule as follows:

```
vchar             = /[\x21-\x7a]/ ;
```

This excludes only the following ASCII printable characters - `{|}~`.   These are all under \x80, so no worries about Unicode on my part.

**Comment by danizen** [comment]
@cKlee , I think I am close now to a workable implementation, and I have now some parameterized tests for accepted patterns and rejected patterns.   However, I am a little unclear how to translate this into an implementation that searches over a Marc record.

I see you have implementations for both perl and php - which would I look to to see how to implement search now that I can parse a string to an AST?   I am working to translate the AST into a "MarcSpec" object, but it may also be possible to search organically by having a Search object parameterized by the record as the target for the parser - the semantics in tatsu's terminology.

**Comment by cKlee** [comment]
> 
> 
> > I see. The PEG Parser uses the ebnf rules. And the problem is that '}' should end parsing the subspec. But the use /[:print]/ entails '}' and thus leeds to the problem?
> 
> Not quite - an honest translation of the ABNF rule:
> 
> ```
> comparisonString  = "\" *VCHAR
> ```
> 
> to tatsu EBNF:
> 
> ```
> comparisonString  = "\" {VCHAR} ;
> ```
> 
> The problem is that `/[:print]/` will match the closing `}` of the surrounding subSpec, and that causes the string not to be matched. If I use instead this rule:
> 
> ```
> comparisonString  = "\" ~ /[a-z0-9A-Z?\-_]*/ ;
> ```
> 
> Then it works, but it is no longer accurate to MarcSpec. It may get me through my project since I likely only need compSpec to match only if the 1st or 2nd indicator is a particular value, but it isn't good enough for you guys to simply include without some notes.
> 
> So, I am now unstuck, but still blocked from creating a proper implementation :)

Very nice approach to use the EBNF like this to parse the string. But it has some drawbacks because its formal and misses the logic.

In my implementations I extract everything between the first '{' and the las '}' and then split by '}{' (with negative lookbehind for escape char '\'). So there is a lot of logic within. Unfortunately code gets very messy and blown up with this approach.

Your solution is absolute fine and comprehensible.

**Comment by cKlee** [comment]
> I see you have implementations for both perl and php - which would I look to to see how to implement search now that I can parse a string to an AST? I am working to translate the AST into a "MarcSpec" object, but it may also be possible to search organically by having a Search object parameterized by the record as the target for the parser - the semantics in tatsu's terminology.

I think I do not understand you properly. What is 'AST'? And do you want to extract data content from a record and therefore you need to search within the record?

**Comment by cKlee** [comment]
> Another question - why not a unaryOperator and binaryOperator rule, e.g. why are `?` and `!` made binary operators anyway:
> 
> ```
> operator          = "=" / "!=" / "~" / "!~" / "!" / "?"
>                     ; equal / unequal / includes / not includes / not exists / exists
> ```

Can you give me an example for a unary operarator? I'm afraid I'm not familiar with this concept :-(

**Comment by danizen** [comment]
So, first status - I can now parse a marcspec pretty generally, a rule like `650[0]$a[0]/1-2$0[0]/2-3` will parse.  The parser generator creates a class `MarcSpecParser` from the EBNF, and this produces an Abstract Syntax Tree (AST).  A subclass forces whitespace tolerance off (see previous comment) and also adds semantics so that the parser builds a `MarcSpec` object.   This object has a `search` method that takes a `pymarc.Record`.

The `MarcSearchParser` does not handle input as general as the `MarcSpecParser`.  Index and charspec that are effective are on the field level, so that `008/0-8` will work, and `650[0]$a$0`, but not the more general form that parses with the pure parser.    The package has its own github.com repository and adequate tests - https://github.com/danizen/pymarcspec/

After implementing this, I wonder whether computers are now fast enough that I should instead convert each record to MARCXML and simply use XPath.  This is a lot of code to write and maintain...

Unary operators are those that only take one argument.  Right now, the way you have written the ABNF, a subspec such as this would be legal - `{$a?$b}`, and I'm not sure what that means.  I would rewrite the ABNF like this:

```
unaryOperator    = "!" / "?"
binaryOperator   = "=" / "!=" / "~" / "!~"
subTermSet       = unaryOperator subTerm / subTerm binaryOperator subTerm
```

This does have the effect of making the operator mandatory - so that `650$a{$9}` would no longer make sense, and instead users would need to be explicit and say `650$a{?$9}`.  Personally, I think explicit operators on the subspec is better.

**Comment by danizen** [comment]
Last statement - I think I may publish to PyPI soon.  I need to do some performance testing and see whether I can use it for the purpose for which I needed a capability of searching MARC.  If it works better than XPath, I will use it - if not, I will deep six all the code as a bad idea for the modern day.
